File: 0948 bag-of-tokens/solution.go
func bagOfTokensScore(tokens []int, power int) int {
    sort.Ints(tokens)
    l,r:=0,len(tokens)-1
    ans,ma:=0,0
    for l<=r {
        if tokens[l] <= power {
            power-=tokens[l]
            l++
            ans++
            ma = max(ma,ans)
        } else if ans > 0 {
            power+=tokens[r]
            r--
            ans--
        } else {
            break
        }
    }
    return ma
}

File: 0948 bag-of-tokens/solution.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int l = 0, r = tokens.length-1;
        int ans = 0,ma = 0;
        while( l<=r ) {
            if(power >= tokens[l]) {
                power-=tokens[l];
                l++;
                ans++;
                ma = Math.max(ma, ans);
            } else if(ans > 0) {
                power+=tokens[r];
                r--;
                ans--;
            } else {
                break;
            }
        }
        return ma;
    }
}
